python/cudf/cudf/_cuda/gpu.pyx

```diff
@@ -307,8 +320,32 @@ def getDeviceCount():
 
     cdef int count
     cdef cudaError_t status = cudaGetDeviceCount(&count)
+
     if status != 0:
-        raise CUDARuntimeError(status)
+        notify_caller_errors = set([
+            cudaErrorInitializationError,
+            cudaErrorInsufficientDriver,
+            cudaErrorInvalidDeviceFunction,
+            cudaErrorNoDevice,
+            cudaErrorInvalidDevice,
+            cudaErrorStartupFailure,
+            cudaErrorInvalidKernelImage,
+            cudaErrorDeviceUninitialized,
+            cudaErrorAlreadyAcquired,
+            cudaErrorOperatingSystem,
+            cudaErrorNotPermitted,
+            cudaErrorNotSupported,
+            cudaErrorSystemNotReady,
+            cudaErrorSystemDriverMismatch,
+            cudaErrorCompatNotSupportedOnDevice,
+            cudaErrorTimeout,
+            cudaErrorUnknown,
+            cudaErrorApiFailureBase
+        ])
+        if status in notify_caller_errors:
+            raise CUDAError(status)
+        else:
+            raise CUDARuntimeError(status)
```

Comment:
Why do we need both `CUDAError` and `CUDARuntimeError` here?

Reply:
Had to introduce a `CUDAError` so that it will get propagated to the user, whereas `CUDARuntimeError` will be caught by the caller in `gpu_utils` and will not be propagated to the user.

Reply:
That sounds like a separation of concerns and that we should make some changes in `gpu_utils` to handle things better.

Reply:
Moved the handling to `gpu_utils`